content/chapters/io/lecture/slides/IPC.md

```diff
@@ -0,0 +1,189 @@
+
+----
+
+### `pipe()` demo
+
+- `demo/IPC/pipe.c`
+
+----
+
+### IPC - Named Pipe (FIFO)
+
+- Bypass unnamed pipes limitations:
+  - Can be used by more than 2 processes
+  - Can be used by unrelated processes
+- Stored on disk as a **special file**
+
+```console
+student@OS:~$ ls -l my_fifo
+prw-r--r-- 1 student student 0 nov 22 18:38 my_fifo
+```
+
+----
+
+### `mkfifo()` demo
+
+- `demo/IPC/fifo.c`
+
+---
+
+### IPC - Socket Interface
+
+![IPC through Socket](./media/IPC-socket.svg)
+
+----
+
+- Designed for interprocess communication
+- Communication between processes on the **same host** or on **different hosts**
+- Supports multiple transport protocols
+  - E.g.: `TCP`, `UDP`
+
+----
+
+### Networking
+
+- Each host is identified by an [**IP address**](https://ocw.cs.pub.ro/courses/uso/laboratoare/laborator-05)
+- For a host to use multiple applications that require network connection we use **ports**
+  - A port is an integet between **0** and **65536**
+- An application is identified by an IP address and a port
+  - E.g.: `217.182.27.243:25565`
+
+----
+
+### Client and Server
+
+- A process that use the socket interface is either a client or a server
+  - The server waits for messages/connections on IP and port
+  - The client attempts to connect to a server using IP and port
+
+----
+
+### TCP and UDP
+
+- The interaction between client and server depends on the protocol they use
+  - `TCP` establishes a reliable connection and ensures all data is transmited
+  - `UDP` sends data faster without checking its integrity
+
+----
+
+### TCP and UDP Simplified
+
+![TCP and UDP simplified](./media/tcp-udp-simplified.png)
+
+---
+
+### TCP Server
+
+----
+
+#### Server - `socket()`
+
+![TCP socket](./media/tcp-socket.svg)
+
+- Create a socket
+
+----
+
+#### Server - `bind()`
+
+![TCP bind](./media/tcp-bind.svg)
+
+- Assign IP and port to the socket
+
+----
+
+#### Server - `listen()`
+
+![TCP listen](./media/tcp-listen.svg)
+
+- Mark the socket passive - it will be used to accept connections
+
+----
+
+#### Server - `accept()`
+
+![TCP accept](./media/tcp-accept.svg)
+
+- Wait for a connection
+- Accept the connection and create a socket for it
+
+---
+
+### TCP Client
+
+----
+
+#### Client - `socket()`
+
+![TCP socket](./media/tcp-socket.svg)
+
+- Create a socket
+
+----
+
+#### Client - `connect()`
+
+![TCP socket](./media/tcp-connect.svg)
+
+- Establish connection with server (IP and port)
+
+---
+
+### TCP - `send()`/`recv()`
+
+![TCP send recv](./media/tcp-send-recv.svg)
+
+- Same as `read()`/`write()` but work only on sockets
+
+----
+
+#### TCP - `send()`
+
+- Use socket to send bytes
+- Return number of sent bytes
+  - **-1** - send failed
+  - **<num** - partial send
+
+----
+
+#### TCP - `recv()`
+
+- Use socket to receive bytes
+- Return number of received bytes
+  - **-1** - recv failed
+  - **0** - connection end (similar to EOF)
+  - **<num** - partial recv
```

Comment:
I would add one `Summary` slide with two columns: `Server` and `Client`. Animated, with the common sequence of actions happening:

1. server creates `socket()`.
1. server `bind()` socket
1. server `listen()` on socket
1. server `accept()` on socket (and blocks)
1. client creates `socket()`
1. (optional for `SOCK_STREAM`) client binds `socket()`
1. client `connect()` to server
1. server `accept()` call returns a new socket
1. server `recv()` request from client
1. client `send()` request to server
1. client `recv()` reply from server
1. server `send()` reply to client
1. client `close()` connection to server
1. server `close()` connection to client